### ISSUE TITLE
Feature : dynamic working set construction

### ISSUE THREAD

**Issue by awilkins** [facebook/nuclide#897]
Having been an active user of Eclipse, I really miss Mylyn when using Atom on codebases with large file counts.

One of the best features of Mylyn is a working set implementation that watches which files you touch and edit, along with hotkeys to allow you to expand hidden bits of the source tree to find new files. Even just with the various methods of jumping to code via references etc, you can build up a relevant working set without thinking about it.

This is also tied into the Mylyn task management system which associates that working set or "context" with a given task, so you can jump from task to task and very quickly get a project tree focussed on just what needs to be done.

With all the building blocks like a tree that supports working sets in place, it would seem like a short hop to having Nuclide support the most important bit of this : watching which files you touch and adding them to an active working set dynamically.

**Comment by shushz** [comment]
Yeah, I wanted to add some dynamic nature to WS definitions. Unfortunately I never got to doing that. Also, it's somewhat tricky to do properly and a lot needs to be considered like:
  - How does one name this WS, we don't have a notion of a currently active task.
  - Several or none WS can be active at any point of time, which one is to be updated?
  - What if there are multiple windows open.
    - They can have some shared projects mounted in them, or not, or one can be a descendant of some mounted in the other. WS are stored in the global configuration, having one window update others may lead to unintuitive UI (which, unfortunately, is already happening in some cases with WS).
  - One natural way of assigning a WS to an activity would be by associating with the source control state (branch, bookmark), but these things are optional, somewhat expensive to recompute, to track and also come with their share of problems.
    - What happens when a user updates to a different changeset in the repository?
    - What if the revision is rebased on top of something else.

Again, though, that is not to say that a dynamically declaring WS is not something that needs to be done, just that given the amount of corner cases to consider it is not something trivial.
